### ISSUE TITLE
Materials do not export well

### ISSUE THREAD

**Issue by quacrobat** [cuddlyogre/ExportLDraw#30]
Hi, great work on this plugin, thank you.

One issue I found is when attempting to export imported LDraw models to glTF (using the default gltf importer/exporter), the colors/materials are not exported or at least are not read by any glTF viewer. According to the Blender glTF IO docs at https://docs.blender.org/manual/en/3.3/addons/import_export/scene_gltf2.html:

> The core material system in glTF supports a metal/rough PBR workflow with the following channels of information:
> 
> Base Color
> Metallic
> Roughness
> Baked Ambient Occlusion
> Normal Map
> Emissive

and

> Tip
> 
> Examining the result of the material import process is a good way to see examples of the types of material nodes and settings that can be exported to glTF.

Would you consider adding something like a "Convert to glTF compatible materials" toggle in import options?

Thanks :)

**Comment by quacrobat** [comment]
And what seems like a very useful video on the subject: 
Creating a glTF-compatible Blender material from a set of PBR textures
https://www.youtube.com/watch?v=70LRKp54zIc

and a more detailed one here:
Highlight: Making a glTF-compatible material from scratch in Blender
https://www.youtube.com/watch?v=YFV33iCXXmo

**Comment by quacrobat** [comment]
hello, any way to fix this?

**Comment by cuddlyogre** [comment]
I'll start looking into this. Real life and other projects have kept me away.

**Comment by cuddlyogre** [comment]
After some cursory testing, it looks like it has something to do with the custom node groups that make up the materials. Exporting with a shader directly connected to output gives me that color, but using a node group gives me all grey.

 I'll keep experimenting to find out for sure.

**Comment by quacrobat** [comment]
thanks for taking the time! happy holidays :)

**Comment by quacrobat** [comment]
> After some cursory testing, it looks like it has something to do with the custom node groups that make up the materials. Exporting with a shader directly connected to output gives me that color, but using a node group gives me all grey.
> 
>  I'll keep experimenting to find out for sure.

yes, perhaps the damage is already done at import time, when the LDraw colors are converted into node materials. An option toggle to import as an alternative node setup may be simpler than trying to change the nodes types at export time. What do you think?

**Comment by cuddlyogre** [comment]
After reading more, this isn’t an uncommon issue and it does appear my material setup is too complex for the exporter. 

I’ll experiment with a way to simplify the node setup.

**Comment by quacrobat** [comment]
Happy New Year!
Did you find time to look into this @cuddlyogre?

**Comment by cuddlyogre** [comment]
The only solution I could find is flattening the material node structure. I'm working on a dynamic programming solution so that I don't have to change the materials.

**Comment by quacrobat** [comment]
Hello, *bump* :)

**Comment by cuddlyogre** [comment]
I haven't forgotten. The different ways I could find to do it haven't really panned out. I'm still working on it though.

**Comment by quacrobat** [comment]
no worries, thanks for your quick reply!

**Comment by quacrobat** [comment]
Congrats on the new release! does it fix the materials export to gltf issue?

**Comment by cuddlyogre** [comment]
Unfortunately no. The gltf exporter doesn't seem to process anything but the principled bsdf shader, so it all still comes out grey. The best I could manage to export is the solid color with some shininess and some transparency. The end result isn't great, so I shelved it until maybe the gltf exporter improved, which is doesn't appear to have.

If you are only looking for basic colors and some specular and maybe transparency, it appears to be doable. If you're looking for materials with anything more than that, I don't think it's doable without having to do a lot of work by hand per part.

**Comment by cuddlyogre** [comment]
I was able to uv unwrap some parts using the sharp edges as a starting point. From there I baked a diffuse and normal texture and exported, set up a material to use them, and the visual results were pretty good. The origin of the object is at 0,0,0 though, so you'd need to sort that out if you're exporting a scene.

![image](https://github.com/cuddlyogre/ExportLDraw/assets/3506795/8f48c54a-86c2-47d5-b363-5482835768a2)

**Comment by quacrobat** [comment]
wow, that's a lot of work, thanks!

**Comment by cuddlyogre** [comment]
I'm still working on making this work as well as I can. The scratches are still in progress. The actual materials themselves are so complicated that making them work with the exporters would involve redoing all of them. The built in Blender bake is rather cumbersome, so I've started using this addon
https://superhivemarket.com/products/simplebake---simple-pbr-and-other-baking-in-blender-2

If you do a Smart UV Project unwrap and use that addon to bake your textures, it works really well.

**Comment by quacrobat** [comment]
Time to let Claude 3.7 have a go at it!
